atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/CoordinationServiceRecorder.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.backup;
+
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.coordination.CoordinationService;
+import com.palantir.atlasdb.coordination.ValueAndBound;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadata;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadataState;
+import com.palantir.atlasdb.internalschema.persistence.CoordinationServices;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+import com.palantir.atlasdb.timelock.api.CompletedBackup;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import java.util.Optional;
+import java.util.function.Function;
+
+final class CoordinationServiceRecorder {
```

Comment:
So how I'd reason through this is at follows:
- Let the backup timestamp be B and the fast forward timestamp be F.
- When we get the state of the coordination service at some timestamp X, this is either empty, or returns a ValueAndBound that has a bound that is at least X. (This performs writes to the coordination service perpetuating the current value to until X.)
- The main property we need here is _sufficient coverage_: every timestamp up to F needs to be known about by the end (the backup consists of values that potentially were written up to just before F, and we need to be able to correctly run clean transactions range). 
- Consistency should follow immediately assuming the coordination service is itself consistent, which we're assuming.

Casework:
- First consider the case where a read of the coordination service happened at F and this returned empty. We must ensure that it is safe to only clean transactions1 in this case. The risk of incorrectness happens if a new transaction schema version is installed at some timestamp, say W, and a transaction starting at X and committing at Y runs and is part of the backup. But this is not possible because W > F.
- Next consider the case where a read happened at F and this returned a map with a validity bound of V. F <= V, by how the coordination service works, and so we have coverage. All transactions in the backup started at a time up to F.

Thus, I'm pretty sure that it is safe to use a single read of coordination at the fast forward timestamp.

Reply:
Thanks for that @jeremyk-91 - I might even steal this and put into an ADR, since we'll be changing how something fundamental works here.